File: gto-support-design/src/main/java/com/google/android/material/tabs/TabLayoutUtils.java
package com.google.android.material.tabs;

import android.graphics.drawable.Drawable;

import java.lang.reflect.Field;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.graphics.drawable.DrawableCompat;
import timber.log.Timber;

public final class TabLayoutUtils {
    private static final String TAG = "TabLayoutUtils";
    private static Field sBaseBackgroundDrawableField;

    @Nullable
    public static Drawable getBackground(@NonNull final TabLayout.Tab tab) {
        return tab.view.getBackground();
    }

    @NonNull
    public static TabLayout.Tab setBackground(@NonNull final TabLayout.Tab tab, @Nullable final Drawable background) {
        try {
            getBaseBackgroundDrawableField().set(tab.view, background);
            tab.view.invalidate();
            tab.parent.invalidate();
        } catch (IllegalAccessException e) {
            Timber.tag(TAG)
                    .d(e, "Error setting the background for a Tab");
        } catch (NoSuchFieldException e) {
            Timber.tag(TAG)
                    .d(e, "Error setting the background for a Tab");
        }

        return tab;
    }

    @NonNull
    public static TabLayout.Tab setBackgroundTint(@NonNull final TabLayout.Tab tab, @ColorInt final int tint) {
        try {
            Drawable bkg = (Drawable) getBaseBackgroundDrawableField().get(tab.view);
            if (bkg != null) {
                bkg = DrawableCompat.wrap(bkg).mutate();
                DrawableCompat.setTint(bkg, tint);
            }
            setBackground(tab, bkg);
        } catch (IllegalAccessException e) {
            Timber.tag(TAG)
                    .d(e, "Error getting the background for a Tab");
        } catch (NoSuchFieldException e) {
            Timber.tag(TAG)
                    .d(e, "Error getting the background for a Tab");
        }

        return tab;
    }

    @NonNull
    public static TabLayout.Tab setVisibility(@NonNull final TabLayout.Tab tab, final int visibility) {
        tab.view.setVisibility(visibility);
        return tab;
    }

    private static Field getBaseBackgroundDrawableField() throws NoSuchFieldException {
        if (sBaseBackgroundDrawableField == null) {
            sBaseBackgroundDrawableField = TabLayout.TabView.class.getDeclaredField("baseBackgroundDrawable");
            sBaseBackgroundDrawableField.setAccessible(true);
        }
        return sBaseBackgroundDrawableField;
    }
}

File: gto-support-design/src/main/java/com/google/android/material/tabs/TabLayoutUtils.kt
@file:JvmName("TabLayoutUtils")

package com.google.android.material.tabs

import android.graphics.drawable.Drawable
import androidx.annotation.ColorInt
import org.ccci.gto.android.common.material.tabs.background
import org.ccci.gto.android.common.material.tabs.notifyPagerAdapterChanged
import org.ccci.gto.android.common.material.tabs.setBackgroundTint

@Deprecated("Since v3.6.1, use extension property from gto-support-material-components instead")
var TabLayout.Tab.background: Drawable?
    get() = background
    set(value) {
        background = value
    }

@Deprecated(
    "Since v3.6.1, use extension method from gto-support-material-components instead",
    ReplaceWith("setBackgroundTint(tint)", "org.ccci.gto.android.common.material.tabs.setBackgroundTint")
)
fun TabLayout.Tab.setBackgroundTint(@ColorInt tint: Int) = setBackgroundTint(tint)

@Deprecated("Since v3.6.1, use extension property from gto-support-material-components instead")
var TabLayout.Tab.visibility: Int
    get() = view.visibility
    set(value) {
        view.visibility = value
    }

@Deprecated(
    "Since v3.6.1, use extension method from gto-support-material-components instead",
    ReplaceWith("notifyPagerAdapterChanged()", "org.ccci.gto.android.common.material.tabs.notifyPagerAdapterChanged")
)
fun TabLayout.notifyPagerAdapterChanged() = notifyPagerAdapterChanged()
